### ISSUE TITLE
Risk of abuse due to lack of simple client-side storage API

### ISSUE THREAD

**Issue by annevk** [WICG/cookie-store#91]
@asutherland raised an interesting point which is that since we lack key-value-based storage in many of the environments this API will be exposed in, it might end up getting abused a lot to get those semantics, resulting in a lot more cookies.

**Comment by asutherland** [comment]
To elaborate a little, my argument is more that Firefox would want to expose something like https://github.com/domenic/async-local-storage (once the layered API standardization foot-gun issue is addressed) before exposing this API because of the risk of misuse in places like ServiceWorkers where LocalStorage is not exposed and IndexedDB is frequently avoided for both dubious and legitimate reasons.  (And ideally all browsers would make a concerted effort to optimize for initial async-local-storage/other get() lookup to make sure the latency of the API is roughly on par with the cookies API much like LocalStorage is today.)

**Comment by pwnall** [comment]
This seems be a cross-API concern regarding how we deploy APIs in browsers. I'm not sure what this specification / explainer could do to address the concern.

FWIW, [the API announcement on developers.google.com](https://developers.google.com/web/updates/2018/09/asynchronous-access-to-http-cookies) starts with "You (probably) don't need cookies". I'd be happy to discuss any concrete suggestions for helping developers choose the right storage APIs for their problems.

**Comment by annevk** [comment]
The problem is that we don't have an alternative API to recommend.

**Comment by annevk** [comment]
I would reopen this, but the way GitHub works prevents that.

**Comment by pwnall** [comment]
I reopened the issue.

It seems to me that "we don't have an alternative API to recommend" is not something that could be solved in this repository. Please help me understand better if I'm wrong.

**Comment by domenic** [comment]
I sympathize with the OP's perspective here, and think this is a genuine issue worth discussing about cookie store. (However, I'll note that @pwnall seems to prefer an issue management style where issues only track things that are fixable by changing spec text, whereas @annevk is assuming the perhaps-more-common issue management style of using the issue tracker to track issues in the space of the proposal.)

I agree that shipping async local storage is an ideal mitigation for this. However, at least for Chrome, I don't think it should be a strict blocker. The timelines don't line up very well; we have web developers in need of the new capabilities this API provides in the cookie space, and we should give it to them as soon as feasible. Whereas shipping async local storage is blocked on both standards and implementation work around making built-in modules feasible. Resolving those things is my day-job P0, but I'd still estimate it'd be at least Q1 2019 before async local storage would be shippable. Whereas from what I understand cookie store is ready to go now-ish.

In that hopefully-short interim period, I think we can do our best to mitigate the problem pointed out by the OP via outreach and education. The spec might be one place for this; developers.google.com is another; you can imagine more. And certainly other browsers might make different tradeoffs, weighing the use cases solved by this proposal vs. the risk of unknowing cookie abuse, which could lead to a different shipping order for the two specs.

Hope this helps!

**Comment by annevk** [comment]
I'm not sure there's a clear suggestion, but IndexedDB is essentially all we have that's not synchronous I/O and available client-side. Though `localStorage` and `sessionStorage` are probably reasonable for small amounts of data at this point, given all the optimizations made.

**Comment by johannhof** [comment]
FWIW I disagree with a lot of the points being made in that blog post and say that (HttpOnly) Cookies are the right place to store session identifiers in most cases, both for security and performance reasons. This article seems to operate under the assumption that all applications exclusively use JavaScript for rendering and fetching information, and that credential theft doesn't exist. Many web applications make the sound decision to rely on server-side rendering and use cookies to obtain session credentials as part of those requests. A better title for this post may have been "Don't use cookies as an arbitrary key-value store in your application", and I think the web has decent alternatives for that:

- Server-side storage with a cookie session identifier
- LocalStorage / SessionStorage for easy storage of small amounts of data
- IndexedDB for more sophisticated storage needs

**Comment by asutherland** [comment]
Quoting @johannhof in https://github.com/WICG/cookie-store/issues/91#issuecomment-2545886898
> A better title for this section may have been "Don't use cookies as an arbitrary key-value store in your application",

Yes, this was the heart of my original concern, especially since at the time the ServiceWorkers WG was seeing a lot of requests that were very much along the lines of requesting access to LocalStorage (and all of the performance problems that could entail but which are largely hidden from users/developers) or potentially unlimited data storage directly on the registration in a structured serialization record (which could have similar performance problems).

It also seemed like https://github.com/WICG/kv-storage was on a viable path to standardization and implementation at that time.  My understanding of the rough consensus of the evolution of that is that we should simply be enhancing IndexedDB to also have a more ergonomic shape along the lines of kv-storage or https://github.com/jakearchibald/idb and I think we are indeed moving that direction once again.

Which is to say, I think the answer to "what should I use for key-value storage in my ServiceWorker" is IndexedDB and we can probably close this.

**Comment by annevk** [comment]
`localStorage` isn't available in many of the environments that the Cookie Store API is available in, so that argument doesn't hold.

I'm not sure if this is new since 2018, but the limits imposed on cookies probably also help make them less attractive as general-purpose storage.

**Comment by asutherland** [comment]
Yes, I think the cookie limits and the resulting risk sites create about expiring their own cookies prematurely makes this less of a concern, especially since I think site authors are much more likely aware of these possible downsides now than they were then.
